File: RaftSwift/Sources/RaftGRPCTransport/GRPCPeerTransport.swift
import RaftCore

/// GRPC implementation of the RaftNodeTransport protocol.
final class GRPCNodeTransport: RaftNodeTransport {
    /// The client pool to use for communication with the server.
    let clientPool: GRPCClientPool

    /// Initializes a new instance of the GRPCNodeTransport class.
    /// - Parameters:
    ///   - clientPool: The client pool to use for communication with the server.
    init(clientPool: GRPCClientPool) {
        self.clientPool = clientPool
    }

    func appendEntries(
        _ request: AppendEntriesRequest,
        to peer: Peer,
        isolation: isolated any Actor,
    ) async throws -> AppendEntriesResponse {
        let client = try await clientPool.client(for: peer)
        let peerClient = Raft_RaftPeer.Client(wrapping: client)

        let response = try await peerClient.appendEntries(.with { grpcRequest in
            grpcRequest.term = UInt64(request.term)
            grpcRequest.leaderID = UInt32(request.leaderID)
            grpcRequest.prevLogIndex = UInt64(request.prevLogIndex)
            grpcRequest.prevLogTerm = UInt64(request.prevLogTerm)
            grpcRequest.entries = request.entries.toGRPC()
            grpcRequest.leaderCommit = UInt64(request.leaderCommit)
        })

        return AppendEntriesResponse(
            term: Int(response.term),
            success: response.success,
        )
    }

    func requestVote(
        _ request: RequestVoteRequest,
        to peer: Peer,
        isolation: isolated any Actor,
    ) async throws -> RequestVoteResponse {
        let client = try await clientPool.client(for: peer)
        let peerClient = Raft_RaftPeer.Client(wrapping: client)

        let response = try await peerClient.requestVote(.with { grpcRequest in
            grpcRequest.term = UInt64(request.term)
            grpcRequest.candidateID = UInt32(request.candidateID)
            grpcRequest.lastLogIndex = UInt64(request.lastLogIndex)
            grpcRequest.lastLogTerm = UInt64(request.lastLogTerm)
        })

        return RequestVoteResponse(
            term: Int(response.term),
            voteGranted: response.voteGranted,
        )
    }

    func installSnapshot(
        _ request: InstallSnapshotRequest,
        on peer: Peer,
        isolation: isolated any Actor,
    ) async throws -> InstallSnapshotResponse {
        let client = try await clientPool.client(for: peer)
        let peerClient = Raft_RaftPeer.Client(wrapping: client)

        let response = try await peerClient.installSnapshot(.with { grpcRequest in
            grpcRequest.term = UInt64(request.term)
            grpcRequest.leaderID = UInt32(request.leaderID)
            grpcRequest.snapshot = request.snapshot.toGRPC()
        })

        return InstallSnapshotResponse(
            term: Int(response.term),
        )
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/transport/grpc/GRPCPeerTransport.kt
package raft.transport.grpc

import raft.core.node.RaftNodeTransport
import raft.core.utils.types.Peer
import raft.core.utils.peer.*
import raft.AppendEntriesRequestKt
import raft.AppendEntriesResponseKt
import raft.transport.grpc.utils.GRPCClientPool
import raft.RaftPeerGrpcKt
import raft.appendEntriesRequest
import raft.installSnapshotRequest
import raft.requestVoteRequest
import raft.transport.grpc.utils.extensions.*

class GRPCNodeTransport(private val clientPool: GRPCClientPool) : RaftNodeTransport {
    override suspend fun appendEntries(
            request: AppendEntriesRequest,
            peer: Peer
    ): AppendEntriesResponse {
        val client = clientPool.clientFor(peer)
        val stub = RaftPeerGrpcKt.RaftPeerCoroutineStub(client)

        val grpcRequest = appendEntriesRequest {
                term = request.term.toLong()
                leaderId = request.leaderId.toInt()
                prevLogIndex = request.prevLogIndex.toLong()
                prevLogTerm = request.prevLogTerm.toLong()
                entries.addAll(request.entries.toGRPC())
                leaderCommit = request.leaderCommit.toLong()
        }

        val grpcResponse = stub.appendEntries(grpcRequest)

        return AppendEntriesResponse(
            term = grpcResponse.term.toInt(),
            success = grpcResponse.success,
        )
    }

    override suspend fun requestVote(request: RequestVoteRequest, peer: Peer): RequestVoteResponse {
        val client = clientPool.clientFor(peer)
        val stub = RaftPeerGrpcKt.RaftPeerCoroutineStub(client)

        val grpcRequest = requestVoteRequest {
            term = request.term.toLong()
            candidateId = request.candidateId
            lastLogIndex = request.lastLogIndex.toLong()
            lastLogTerm = request.lastLogTerm.toLong()
        }

        val response = stub.requestVote(grpcRequest)

        return RequestVoteResponse(
            term = response.term.toInt(),
            voteGranted = response.voteGranted,
        )
    }

    override suspend fun installSnapshot(
            request: InstallSnapshotRequest,
            peer: Peer
    ): InstallSnapshotResponse {
        val client = clientPool.clientFor(peer)
        val stub = RaftPeerGrpcKt.RaftPeerCoroutineStub(client)

        val grpcRequest = installSnapshotRequest {
            term = request.term.toLong()
            leaderId = request.leaderId
            snapshot = request.snapshot.toGRPC()
        }

        val grpcResponse = stub.installSnapshot(grpcRequest)

        return InstallSnapshotResponse(
            term = grpcResponse.term.toInt(),
        )
    }
}
